packages/build/src/plugins_core/blobs_upload/utils.ts

```diff
@@ -0,0 +1,61 @@
+import fs from 'node:fs'
+import fsPromises from 'node:fs/promises'
+import path from 'node:path'
+
+import { fdir } from 'fdir'
+
+const metadataPrefix = '$'
+const metadataSuffix = '.json'
+
+export const getBlobsDir = function ({ buildDir, publishDir }) {
+  const blobsPath = '.netlify/blobs/deploy'
+  const absoluteBlobsPath = path.resolve(buildDir, publishDir, blobsPath)
+  const blobsDir = path.normalize(absoluteBlobsPath)
+  return blobsDir
+}
+
+export const anyBlobsToUpload = function ({ buildDir, publishDir }) {
+  const blobsDir = getBlobsDir({ buildDir, publishDir })
+  const { files } = new fdir().onlyCounts().crawl(blobsDir).sync()
+  return files > 0
+}
+
+// Given output directory, find all file paths to upload excluding metadata files
+export async function getKeysToUpload(blobsDir): Promise<string[]> {
+  const files = await new fdir()
+    .withRelativePaths() // we want the relative path from the blobsDir
+    .filter((fpath) => !path.basename(fpath).startsWith(metadataPrefix))
+    .crawl(blobsDir)
+    .withPromise()
+
+  // normalize the path separators to all use the forward slash
+  return files.map((f) => f.split(path.sep).join('/'))
+}
+
+// Read a file and its metadata from the blobs directory
+export async function getFileWithMetadata(
+  blobsDir: string,
+  key: string,
+): Promise<{ data: Buffer; metadata: Record<string, string> }> {
+  const contentPath = path.join(blobsDir, key)
+  const dirname = path.dirname(key)
+  const basename = path.basename(key)
+  const metadataPath = path.join(blobsDir, dirname, `${metadataPrefix}${basename}${metadataSuffix}`)
+
+  const [data, metadata] = await Promise.all([fsPromises.readFile(contentPath), readMetadata(metadataPath)]).catch(
+    (err) => {
+      throw new Error(`Failed while reading '${key}' and its metadata: ${err.message}`)
+    },
+  )
+
+  return { data, metadata }
+}
+
+async function readMetadata(metadataPath: string): Promise<Record<string, string>> {
+  let metadata = {}
+  if (fs.existsSync(metadataPath)) {
```

Comment:
Checking if the file exists before trying to read it is usually [not a recommended pattern](https://nodejs.org/api/fs.html#fsexistspath-callback). I would recommend doing the `readFile` in a `try/catch` and looking for `error.code === "ENOENT"` as signal that the file doesn't exist.

Reply:
Updated in f90332636